### PR TITLE
modify r_min

### PR DESCRIPTION
**Pull request by SYEDZIAULISLAM**
if (r_min < 0)
    {
        throw std::invalid_argument("RDF requires r_min to be non_negetive.");
    }

### CODE REVIEW

File: cpp/density/RDF.cc [hunk 1]
```diff
@@ -22,6 +22,10 @@ RDF::RDF(unsigned int bins, float r_max, float r_min, bool normalize)
     {
         throw std::invalid_argument("RDF requires r_max to be positive.");
     }
+    if (r_min < 0)
+    {
+        throw std::invalid_argument("RDF requires r_min to be zero or positive.");
```
Comment (bdice) [inline cpp/density/RDF.cc]: We use the term "non-negative" in other parts of the code. Let's keep consistent with that naming convention. Examples below.
```suggestion
        throw std::invalid_argument("RDF requires r_min to be non-negative.");
```

https://github.com/glotzerlab/freud/blob/eb2a97bac302ed45236f26e0a7e6a0346071c579/cpp/locality/PeriodicBuffer.cc#L20

https://github.com/glotzerlab/freud/blob/eb2a97bac302ed45236f26e0a7e6a0346071c579/cpp/order/SolidLiquid.cc#L17-L19

**Review by bdice — COMMENTED** [review]
Hi @SYEDZIAULISLAM, thanks for your interest in helping solve issue #821! Could you copy-paste the contents of [this PR template](https://github.com/glotzerlab/freud/blob/master/.github/PULL_REQUEST_TEMPLATE.md) into the PR description, and fill out each section? It helps us track the status of each pull request. Thank you!

File: cpp/density/RDF.cc (hunk 1 above)
Comment (SYEDZIAULISLAM) [reply cpp/density/RDF.cc]: okay

**Review by SYEDZIAULISLAM — COMMENTED** [review]
modify **r_min** and import all the library which are not supported by wheels in previous test cases

**Comment by SYEDZIAULISLAM** [comment]
@bdice can u look , what the error that is coming?

File: cpp/density/RDF.h [hunk 2]
```diff
@@ -17,7 +17,24 @@ class RDF : public locality::BondHistogramCompute
 {
 public:
     //! Constructor
-    RDF(unsigned int bins, float r_max, float r_min = 0, bool normalize = false);
+    RDF(unsigned int bins, float r_max, float r_min = 0, bool normalize = false){
```
Comment (bdice) [inline cpp/density/RDF.h:23]: The constructor is implemented in `RDF.cc`. The header file should only contain the declaration of the constructor and its arguments. Please revert these changes in `RDF.h` and add the block for `if (r_min < 0)` to the `RDF.cc` file instead.

File: cpp/locality/AABBQuery.cc [hunk 3]
```diff
@@ -66,6 +66,13 @@ void AABBIterator::updateImageVectors(float r_max, bool _check_r_max)
     box::Box box = m_neighbor_query->getBox();
     vec3<float> nearest_plane_distance = box.getNearestPlaneDistance();
     vec3<bool> periodic = box.getPeriodic();
+
+    
+        if (r_max <= 0)
+    {
+        throw std::invalid_argument("RDF requires r_max to be positive.");
+    }
```
Comment (bdice) [inline cpp/locality/AABBQuery.cc]: The `r_max` and `r_min` parameters should be validated in the parent class `NeighborQueryPerPointIterator`, not in this method. Validation should have already occurred before reaching this point in the code.

Revert this change and add validation logic to the empty constructor body here instead:
https://github.com/glotzerlab/freud/blob/eb2a97bac302ed45236f26e0a7e6a0346071c579/cpp/locality/NeighborQuery.h#L262-L268

edit: It looks like you have already added validation logic to that constructor. 👍

File: cpp/locality/AABBQuery.h [hunk 4]
```diff
@@ -123,7 +123,20 @@ class AABBIterator : public NeighborQueryPerPointIterator
         : NeighborQueryPerPointIterator(neighbor_query, query_point, query_point_idx, r_max, r_min,
                                         exclude_ii),
           m_aabb_query(neighbor_query)
-    {}
+    {
```
Comment (bdice) [inline cpp/locality/AABBQuery.h]: See previous comment -- revert this change and apply the validation logic in the `NeighborQueryPerPointIterator` [constructor](https://github.com/glotzerlab/freud/blob/eb2a97bac302ed45236f26e0a7e6a0346071c579/cpp/locality/NeighborQuery.h#L262-L268) instead. I think you can safely revert all the changes in this file and `LinkCell.h`.

File: cpp/locality/NeighborList.cc [hunk 5]
```diff
@@ -163,6 +163,18 @@ template unsigned int NeighborList::filter(bool*);
 
 unsigned int NeighborList::filter_r(float r_max, float r_min)
 {
+    if (r_max <= 0) 
+ { 
+     throw std::invalid_argument("RDF requires r_max to be positive.");
```
Comment (bdice) [inline cpp/locality/NeighborList.cc]: Update the names of the errors:
```suggestion
     throw std::invalid_argument("NeighborList.filter_r requires r_max to be positive."); 
```

File: cpp/locality/NeighborList.h [hunk 6]
```diff
@@ -112,7 +112,21 @@ class NeighborList
     template<typename Iterator> unsigned int filter(Iterator begin);
     //! Remove bonds in this object based on minimum and maximum distance
     //  constraints. Returns the number of bonds removed.
-    unsigned int filter_r(float r_max, float r_min = 0);
+    unsigned int filter_r(float r_max, float r_min = 0){
```
Comment (bdice) [inline cpp/locality/NeighborList.h:118]: This is the header file. The implementation of this method is in `NeighborList.cc`, and the validation logic has been added there already. Please revert this change, since only the method declaration must occur in the header.

File: cpp/locality/NeighborQuery.h [hunk 7]
```diff
@@ -265,7 +265,20 @@ class NeighborQueryPerPointIterator : public NeighborPerPointIterator
         : NeighborPerPointIterator(query_point_idx), m_neighbor_query(neighbor_query),
           m_query_point(query_point), m_finished(false), m_r_max(r_max), m_r_min(r_min),
           m_exclude_ii(exclude_ii)
-    {}
+    {
+        if (r_max <= 0)
+    {
+        throw std::invalid_argument("RDF requires r_max to be positive.");
```
Comment (bdice) [inline cpp/locality/NeighborQuery.h]: This is the correct place to implement the validation logic for the `NeighborQuery` classes like `AABBQuery` and `LinkCell`. 👍  There are a couple changes needed.
1. Change the message:
```suggestion
        throw std::invalid_argument("Neighbor queries require r_max to be positive.");
```
2. Fix the indentation. You can run `clang-format` on the repository or just look at the rest of the code to learn the indentation style. See the developer documentation for info: https://freud.readthedocs.io/en/latest/reference/development/howtoadd.html

**Review by bdice — CHANGES_REQUESTED** [review]
Hi @SYEDZIAULISLAM, thanks for your work on this! I have a few changes to request. The CI failures are from a failure to compile, due to the changes you made in header files (`*.h`) that should be in the implementation files (`*.cc`) instead. If you revert some of the changes that I point out in this review, it should fix the build error. In case you haven't seen it, we also have a [Developer Guide](https://freud.readthedocs.io/en/latest/reference/development.html) discussing the project's code style and organization.
